src/Xamarin.Android.Build.Tasks/Utilities/TypeMappingDebugNativeAssemblyGeneratorCLR.cs

```diff
@@ -0,0 +1,251 @@
+using System;
+using System.Collections.Generic;
+
+using Microsoft.Build.Utilities;
+
+using Xamarin.Android.Tasks.LLVMIR;
+
+namespace Xamarin.Android.Tasks;
+
+class TypeMappingDebugNativeAssemblyGeneratorCLR : LlvmIrComposer
+{
+	const string JavaToManagedSymbol = "map_java_to_managed";
+	const string ManagedToJavaSymbol = "map_managed_to_java";
+
+	// These names MUST match src/native/clr/include/xamarin-app.hh
+	const string TypeMapSymbol = "type_map";
+	const string UniqueAssembliesSymbol = "type_map_unique_assemblies";
+	const string AssemblyNamesBlobSymbol = "type_map_assembly_names_blob";
+
+	sealed class TypeMapContextDataProvider : NativeAssemblerStructContextDataProvider
+	{
+		public override ulong GetBufferSize (object data, string fieldName)
+		{
+			var map_module = EnsureType<TypeMap> (data);
+			if (String.Compare ("java_to_managed", fieldName, StringComparison.Ordinal) == 0 ||
+			    String.Compare ("managed_to_java", fieldName, StringComparison.Ordinal) == 0) {
+				    return map_module.entry_count;
+			    }
```

Comment:
Indentation here is weird:

  * Tabs vs. spaces
  * this `}` should have the same indent as it's `if`.
  * The body of the `if` expression should be indented *more* than the `if` body, not less.

```csharp
if (String.Compare ("java_to_managed", fieldName, StringComparison.Ordinal) == 0 ||
        String.Compare ("managed_to_java", fieldName, StringComparison.Ordinal) == 0) {
    return map_module.entry_count;
}
```